subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+
+	if ((zms_ate_crc8_check(entry)) || (fs->sector_cycle != entry->cycle_cnt)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/* zms_ate_valid_different_sector validates an ate that is in a different
+ * sector than the active one. It takes as argument the cycle_cnt of the
+ * sector where the ATE to be validated is stored
+ *     return 1 if crc8 and cycle_cnt are valid,
+ *            0 otherwise
+ */
+static int zms_ate_valid_different_sector(struct zms_fs *fs, const struct zms_ate *entry,
+					  uint8_t cycle_cnt)
+{
+	size_t ate_size;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	if ((zms_ate_crc8_check(entry)) || (cycle_cnt != entry->cycle_cnt)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+static inline int zms_get_cycle_on_sector_change(struct zms_fs *fs, uint64_t addr,
+						 int previous_sector_num, uint8_t *cycle_cnt)
+{
+	int rc;
+
+	/* read the ate cycle only when we change the sector
+	 * or if it is the first read
+	 */
+	if ((SECTOR_NUM(addr) != previous_sector_num) || (previous_sector_num == -1)) {
+		rc = zms_get_sector_cycle(fs, addr, cycle_cnt);
+		if (rc == -ENOENT) {
+			/* sector never used */
+			*cycle_cnt = 0;
+		} else if (rc) {
+			/* bad flash read */
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/* zms_close_ate_valid validates an sector close ate: a valid sector close ate:
+ * - valid ate
+ * - len = 0 and id = ZMS_HEAD_ID
+ * - offset points to location at ate multiple from sector size
+ * return 1 if valid, 0 otherwise
+ */
+static int zms_close_ate_valid(struct zms_fs *fs, const struct zms_ate *entry)
+{
+	size_t ate_size;
+
+	if ((!zms_ate_valid_different_sector(fs, entry, entry->cycle_cnt)) || (entry->len != 0U) ||
+	    (entry->id != ZMS_HEAD_ID)) {
+		return 0;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	if ((fs->sector_size - entry->offset) % ate_size) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/* zms_empty_ate_valid validates an sector empty ate: a valid sector empty ate:
+ * - valid ate
+ * - len = 0xffff and id = 0xffffffff
+ * return 1 if valid, 0 otherwise
+ */
+static int zms_empty_ate_valid(struct zms_fs *fs, const struct zms_ate *entry)
+{
+	if ((!zms_ate_valid_different_sector(fs, entry, entry->cycle_cnt)) ||
+	    (entry->len != 0xffff) || (entry->id != ZMS_HEAD_ID)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/* store an entry in flash */
+static int zms_flash_write_entry(struct zms_fs *fs, uint32_t id, const void *data, size_t len)
+{
+	int rc;
+	struct zms_ate entry;
+
+	entry.id = id;
+	entry.len = (uint16_t)len;
+	entry.cycle_cnt = fs->sector_cycle;
+
+	if (len > 8) {
+		/* only compute CRC if len is greater than 8 bytes */
+		if (IS_ENABLED(CONFIG_ZMS_DATA_CRC)) {
+			entry.data_crc = crc32_ieee(data, len);
+		}
+		entry.offset = (uint32_t)(fs->data_wra & ADDR_OFFS_MASK);
+	} else if ((len > 0) && (len <= 8)) {
+		/* Copy data into entry for small data ( < 8B) */
+		uint8_t *data8 = (uint8_t *)data;
+
+		memcpy(&entry.data, data8, len);
+	}
+
+	zms_ate_crc8_update(&entry);
```

Comment:
You have to zero the entire `entry` before calculating the CRC given that not all members are necessarily initialized (e.g. when writing a value smaller than 8 bytes (is there a unit test for this case?))

Reply:
No there is no unitary test covering this. I will memset the data member. That's the only member non intialized here

Reply:
If you only zero particular members, you also need to zero `entry.data_crc` if data CRC is disabled.